type-metadata/src/lib.rs

```diff
@@ -1,7 +1,346 @@
-#[cfg(test)]
-mod tests {
-	#[test]
-	fn it_works() {
-		assert_eq!(2 + 2, 4);
+// Copyright 2019 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod registry;
+pub use registry::Registry;
+
+mod tests;
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub enum IdentKind {
+	Custom(CustomIdent),
+
+	// primitives, including build-in types and common precludes
+	Bool,
+	Str,
+	U8,
+	U16,
+	U32,
+	U64,
+	U128,
+	I8,
+	I16,
+	I32,
+	I64,
+	I128,
+	Array(ArrayIdent),
+	Slice(SliceIdent),
+	Tuple(TupleIdent),
+	Option(OptionIdent),
+	Result(ResultIdent),
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct Namespace(Vec<&'static str>);
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct CustomIdent {
+	pub name: &'static str,
+	pub namespace: Namespace,
+	pub type_params: Vec<IdentKind>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ArrayIdent {
+	pub len: u16,
+	pub type_param: Box<IdentKind>,
+}
+impl ArrayIdent {
+	fn new(len: u16, type_param: IdentKind) -> Self {
+		ArrayIdent {
+			len,
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TupleIdent {
+	pub type_params: Option<Vec<IdentKind>>,
+}
+impl TupleIdent {
+	fn new(type_params: Option<Vec<IdentKind>>) -> Self {
+		TupleIdent { type_params }
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct SliceIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl SliceIdent {
+	fn new(type_param: IdentKind) -> Self {
+		SliceIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct OptionIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl OptionIdent {
+	fn new(type_param: IdentKind) -> Self {
+		OptionIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ResultIdent {
+	pub type_params: (Box<IdentKind>, Box<IdentKind>),
+}
+impl ResultIdent {
+	fn new(type_params: (IdentKind, IdentKind)) -> Self {
+		ResultIdent {
+			type_params: (Box::new(type_params.0), Box::new(type_params.1)),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum TypeDef {
+	Primitive,
```

Comment:
I just saw that my above proposal for introducing `PrimitiveIdent` would be a perfect fit with this.

Reply:
`Primitive` really means no type def provided (may be call it something else?). It will also be used for types that we cannot describe encoding format. e.g. [`Address`](https://github.com/paritytech/substrate/blob/master/srml/indices/src/address.rs#L87)

Reply:
Then, how about an `Option<TypeDef>` ? :D

Reply:
That will work, except again required one more byte with naive binary encoding format. I want to avoid to have custom parity-codec implementation because it just means more code (in both Rust and JS and all other future SDKs side)

Reply:
Then maybe just call the variant `None`: `TypeDef::None`

Reply:
Cool, renamed to `TypeDef::None`.